packages/labs/context/src/lib/controllers/context-consumer.ts

```diff
@@ -0,0 +1,74 @@
+/**
+ * @license
+ * Copyright 2021 Google LLC
+ * SPDX-License-Identifier: BSD-3-Clause
+ */
+
+import {ContextEvent} from '../context-event.js';
+import {Context, ContextType} from '../create-context.js';
+import {ReactiveController, ReactiveElement} from 'lit';
+
+/**
+ * ContextConsumer is a ReactiveController which binds a custom-element's
+ * lifecycle to the Context API. When an element is connected to the DOM it
+ * will emit the context-request event, invoking the callback set on the
+ * controller when the context request is satisfied. It will also call
+ * the dispose method provided by the Context API when the element is
+ * disconnected.
+ */
+export class ContextConsumer<
+  Ctx extends Context<unknown>,
+  HostElement extends ReactiveElement
+> implements ReactiveController
+{
+  private provided = false;
+
+  constructor(
+    protected host: HostElement,
+    private context: Ctx,
+    private callback: (value: ContextType<Ctx>, dispose?: () => void) => void,
+    private multiple: boolean = false
+  ) {
+    this.host.addController(this);
+  }
+
+  private dispose?: () => void;
+
+  hostConnected(): void {
+    this.host.dispatchEvent(
```

Comment:
@benjamind 

In my port of this library [here](https://github.com/holochain-open-dev/context/tree/main/ui), I've had to add a `setTimeout` here because of an edge case:

- I have a parent element that is providing context to slotted elements:
```js
export class SynSession extends ScopedElementsMixin(LitElement) {
...
    render() {
    return html`
    <slot
      ${provide(
        synSessionContext,
        this.synStore?.sessionStore(this.sessionHash)
      )}
    ></slot>`;
  }

```

and then the context consumer is like: 

```js
export class SynFolks extends ScopedElementsMixin(LitElement) {
  @contextProvided({ context: synSessionContext, multiple: true })
  session!: SessionStore<any, any>;

  render() {
    if (!this.session) return html`There is no active session`;

    return html` Hi!`;
  }
}
```

and the html is like:

```html
  <syn-context>
    <syn-folks />
  </syn-context>
```

What happens is that the context consumer element gets connected before the provider does, and thus the event gets fired but not listened to. For now, the `setTimeout` workaround works, but I know it's not the most elegant solution.

Reply:
I handled this use case in this library: https://github.com/blikblum/wc-context

It tracks when a context request is not handled and re send the event after a microtask

Reply:
Elegant solutions, yeah we should work something like this in I guess. Thanks for the feedback!

Reply:
I can see this problem happening with the `provide()` directive, since it's not set up until rendering, which is at least a microtask after connected. but it shouldn't happen with ContextProvider since the listener is set up on connected. This could be a reason to recommend the controller over the directive. Perhaps the controller could take a selector to scope it to cover the subtree cases of the directive.

Reply:
FYI this occurs when provider is in a shadow Dom host and the consumer is slotted even when "providing" in connected callback.

The slotted element connected callback is fired before the  host connected callback.

Reply:
I also could fork this and play around. It took time to properly setup lit@2.0.0-rc  but it's working.

(I'm aware this is a lit implementation, but...)

say I have a team who will use the components, and I want them to be free to use plain JavaScript, or their framework of choice.

I wonder how we could make a provider builerplate that isn't bound to lit.
 I'll be spending time trying to make a Provider HTMLElement and probably see to have methods for updating fed in as argument.

Reply:
I have added the concept of a `ContextRoot` to this implementation that does something similar to these. It listens at its attached node for unhandled context request events, and also introduces the concept that providers should advertise themselves via a `context-provider` event. Once it sees a provider for an unhandled context request it re-dispatches the event at the source. Seems to work nicely without being too complex or needing to walk the DOM tree.